proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,374 @@
+
+### Objective-C
+
+Objective-C uses C structs for its equivalent of fixed-contents structs, but *non-*fixed-contents structs are instead represented using classes, often immutable classes. (An immutable class sidesteps any questions about value semantics as long as you don't test its identity, e.g. with `===`.) The presence of these classes in Apple's frameworks demonstrates the need for non-fixed-contents structs in Swift.
+
+Old versions of Objective-C (as in, older than iOS) had the same restrictions for the instance variables of a class as C does for the fields of a struct: any change would result in a change in layout, which would break any code that accessed instance variables directly. This was largely acceptable for classes that didn't need to be subclassed, but those that did had to keep their set of instance variables fixed. (This led to the same sort of "reserved" tricks used by C structs.) This restriction was lifted at the cost of some first-use set-up time with Apple's "non-fragile Objective-C runtime", introduced with iOS and with the 64-bit version of Mac OS X; it is now also supported in the open-source GNUstep runtime.
+
+
+### Other languages
+
+Many other languages allow defining structs, but most of them either don't bother to define a stable ABI or don't allow modifying the fields of a struct once it becomes part of an ABI. Haskell and Ocaml fall into this bucket.
+
+
+## Source compatibility
+
+This change does not impact source compatibility. It has always been a source-compatible change to modify a struct's non-public stored instance properties, and to add a new public stored instance property, as long as the struct's existing API does not change (including any initializers).
+
+
+## Effect on ABI stability
+
+Currently, the layout of a public struct is known at compile time in both the defining library and in its clients. For a library concerned about binary compatibility, the layout of a non-fixed-contents struct must not be exposed to clients, since the library may choose to add new stored instance properties that do not fit in that layout in its next release, or even remove existing properties as long as they are not public.
+
+These considerations should not affect libraries shipped with their clients, including SwiftPM packages. These libraries should always have binary stability mode turned off, indicating that the compiler is free to optimize based on the layout of a struct (because the library won't change).
+
+
+## Effect on Library Evolution
+
+The set of binary-compatible changes to a struct's stored instance properties is described above.
+
+Taking an existing struct and marking it `@frozen` is something we'd like to support without breaking binary compatibility, but there is no design for that yet.
+
+Removing `@frozen` from a struct is not allowed; this would break any existing clients that rely on the struct's existing layout.
```

Comment:
Enums need to be discussed here too now, although you can probably lift that from the early versions of SE-0192.

Reply:
I think "type" works here with this level of detail.